### PR TITLE
Extend bpf verifier tests to test socket programs

### LINKED ISSUE #8424: Add cgroup verifier tests to CI
The CI runs `test/bpf/verifier-test.sh` automatically.

Recently, structure was added to support testing the cgroups BPF programs via this script, however for now it just skips running the test:

https://github.com/cilium/cilium/blob/32daf1eb19bb13bf4c8ea02e2844261df97a7b31/test/bpf/verifier-test.sh#L66-L70

We should extend this so that it actually attempts to attach these cgroups programs and checks that they pass the verifier.

This can be based loosely on the implementation in `bpf/init.sh`:
https://github.com/cilium/cilium/blob/32daf1eb19bb13bf4c8ea02e2844261df97a7b31/bpf/init.sh#L335
https://github.com/cilium/cilium/blob/32daf1eb19bb13bf4c8ea02e2844261df97a7b31/bpf/init.sh#L342

Caveats:
* On older kernels, these programs might not attach due to a lack of kernel support. Similarly to the XDP prog tests in the script, we would need to detect this lack of support and skip running these tests if there's no support for cgroups programs.

### PR DESCRIPTION
**Pull request by joestringer**
This PR makes various cleanups in the `test/bpf/verifier-test.sh` script, adds a new tool `tools/maptool` for creating `sockmap` and `eppolicymap` map times from a simple CLI, and finally adds support for loading the socket-level BPF programs into the verifier test script.

The PR is spread out across several commits to allow review commit-by-commit.

Note, the script is only run in the CI under kernel 4.9.160, which doesn't have support for any of the socket-level functionality. However I've written the script in such a way that it detects available support before running the tests (with an override option). As such, the CI will *not* run this new functionality; that said, with it in the tree we will have the option to do so in future, and furthermore developers can manually run it in their own environments on a newer kernel.

Fixes: #8424

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/9339)

### CODE REVIEW

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

**Comment by joestringer** [comment]
test-me-please

File: pkg/maps/eppolicymap/eppolicymap.go [hunk 1]
```diff
@@ -79,10 +72,29 @@ func CreateEPPolicyMap() {
...
 
-	if _, err := EpPolicyMap.OpenOrCreate(); err != nil {
+	_, err := EpPolicyMap.OpenOrCreate()
+	return err
+}
+
+// CreateEPPolicyMap will create both the innerMap (needed for map in map types) and
+// then after BPFFS is mounted create the epPolicyMap. We only create the innerFd once
+// to avoid having multiple inner maps.
+func CreateEPPolicyMap() {
```
Comment (aanm) [inline pkg/maps/eppolicymap/eppolicymap.go:96]: Maybe we should execute this in an `init()` function?

File: pkg/maps/sockmap/sockmap.go [hunk 2]
```diff
@@ -100,20 +101,36 @@ const (
...
+	})
+
+	_, err := SockMap.OpenOrCreate()
+	return err
+}
+
 // SockmapCreate will create sockmap map
 func SockmapCreate() {
-	SockMap.OpenOrCreate()
+	if err := CreateWithName(mapName); err != nil {
```
Comment (aanm) [inline pkg/maps/sockmap/sockmap.go:133]: same comment wrt `init()`

**Review by aanm — COMMENTED** [review]
Non blocking from my side just a small comment.

File: pkg/maps/eppolicymap/eppolicymap.go [hunk 3]
```diff
@@ -36,7 +36,7 @@ var (
 
 const (
 	// MaxEntries represents the maximum number of endpoints in the map
-	MaxEntries = 65535
+	MaxEntries = 65536
```
Comment (brb) [inline pkg/maps/eppolicymap/eppolicymap.go:39]: Any implications during the upgrade due to the bump?

File: pkg/maps/eppolicymap/eppolicymap.go (hunk 1 above)
Comment (joestringer) [reply pkg/maps/eppolicymap/eppolicymap.go:96]: Map creation during init causes problems with our BPFFS path setup doesn't it?

Anyway, both of these maps are created depending on kernel version and Cilium settings. On older kernels we can't (or don't need to) create either of them.

File: pkg/maps/eppolicymap/eppolicymap.go (hunk 3 above)
Comment (joestringer) [reply pkg/maps/eppolicymap/eppolicymap.go:39]: Great question. Let me explore this a bit..

Existing `sockmap` programs will continue to reference the old `eppolicymap` and each entry in the old map holds a reference to the map that the `eppolicymap` points to. So lookups coming in before endpoints are regenerated will continue to work, even when the map is recreated/pinned on the filesystem.

So what happens next? Well, we start regenerating endpoints. For each endpoint that we regenerate, it triggers update of the new `eppolicymap` map with the endpoint's `policymap` fd so that socket programs will be able to reference that endpoint. Note however that until we (re)load the sockops programs, nothing refers to the new `eppolicymap`.

I'm not sure if there is any ordering between the (re)load of socket programs vs. all endpoints regeneration, but let's say that these happen at the same time. When the socket progs are loaded, they may reference the new map, which doesn't have links to all local endpoint `policymap` fds. This means that sockmap progs will not be able to find the policy (including proxy redirects) from arbitrary endpoints. However the only code in the `bpf/sockops/bpf_sockops.c` that relies upon this is the code which updates the `sock_ops_map` to support redirects to proxy sockets; During upgrade, proxy sockets are disrupted anyway because we redeploy the entire proxy.

Furthermore, sockmap is not yet out of beta so I don't know if upgrade considerations are paramount yet.

In conclusion, no, there should be no implications during upgrade.

Comment (jrfastab) [reply pkg/maps/eppolicymap/eppolicymap.go:39]: Also if we fail the lookup in the ENDPOINTS_MAP then we will just pass the connection onto the stack and worse case sockops will not be used.

Comment (joestringer) [reply pkg/maps/eppolicymap/eppolicymap.go:39]: Ah right, I missed the redirect program. But yes, that will just fall back to the packet path (and the BPF enforcement that occurs there).

Comment (joestringer) [reply pkg/maps/eppolicymap/eppolicymap.go:39]: I think there's actually still a bit more subtlety here where this would effectively turn off sockmap redirection for existing connections during upgrade. This could have some follow-on effects but it's kinda hard to say for sure how it will interact.

I guess this suggests that we should consider setting up a variation on our upgrade tests where we run with sockmap ;)

That said I'm still not too worried given the current stability status of sockmap.

**Review by brb — APPROVED** [review]

**Comment by joestringer** [comment]
test-me-please
